### ISSUE TITLE
Constructor Functions

### ISSUE THREAD

**Issue by carver** [ethereum/web3.py#666]
Continuation of #577 

Add constructor method to replace and expand on the current `contract.deploy()` capability. Proposed API:

- `contract.constructor(*args, **kwargs).estimateGas()`
- `contract.constructor(*args, **kwargs).transact()`
- `contract.constructor(*args, **kwargs).buildTransaction()`

**Comment by carver** [comment]
@dangell7 started in on this with #595 -- are you interested in continuing the work? You should be unblocked now that the `chainID` issue is fixed. I think we can get another bounty on just this piece, since the fallback work in #577 was more than expected.

**Comment by pipermerriam** [comment]
confirmed that I'll bounty this issue with an equivalent ETH amount as the previous one.

**Comment by carver** [comment]
How about you @zheli - are you interested and available to work on this? (given that an equivalent bounty is coming soon)

**Comment by zheli** [comment]
I would be happy to work on this issue :)

**Comment by pipermerriam** [comment]
@zheli still waiting for txn to confirm, but i just attached a `0.4` eth bounty to this.  We'd like to get this closed out quick so if you're available to work on it and want to do it.... go for it!

**Comment by dangell7** [comment]
Yeah sorry about going mia. I’ve been busy, the team switched to .js 😔but it really should just need to be tested. I’ve been using the forked repo for awhile now. You’re welcome to copy anything I wrote.

**Comment by pipermerriam** [comment]
I used too low a gas price yesterday evening, trying with a higher one.  Either way, this issue will have a bounty.

**Comment by gitcoinbot** [comment]
__This issue now has a funding of 0.4 ETH (346.49 USD) attached to it.__

 * If you would like to work on this issue you can claim it [here](https://gitcoin.co/funding/details?url=https://github.com/ethereum/web3.py/issues/666).
 * If you've completed this issue and want to claim the bounty you can do so [here](https://gitcoin.co/funding/details?url=https://github.com/ethereum/web3.py/issues/666)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $7705.12 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by gitcoinbot** [comment]
__Work has been started on the 0.4 ETH (346.49 USD) funding by__: 
 1. [@zheli](https://gitcoin.co/profile/zheli) 

 __Please work together__ and coordinate delivery of the issue scope. Gitcoin doesn't know enough about everyones skillsets / free time to say who should work on what, but we trust that the community is smart and well-intentioned enough to work together.  As a general rule; if you start work first, youll be at the top of the above list ^^, and should have 'dibs' as long as you follow through. 

 On the above list? Please leave a comment to let the funder (@pipermerriam) and the other parties involved what you're working, with respect to this issue and your plans to resolve it.  If you don't leave a comment, the funder may expire your submission at their discretion. 

 * Learn more [on the gitcoin issue page](https://gitcoin.co/funding/details?url=https://github.com/ethereum/web3.py/issues/666)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $7705.12 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by zheli** [comment]
Thx @pipermerriam! I will probably make a new pull request based on the changes in #595.

**Comment by nueverest** [comment]
Would love to see this implemented.

**Comment by vs77bb** [comment]
Hi @zheli Vivek from Gitcoin here checking in on behalf of Gitcoin Bot, who we're still training 🙂 any update here? Hope you all are doing well! @pipermerriam @carver

**Comment by zheli** [comment]
Hi @vs77bb sorry I was a bit busy with my daytime job in the past two weeks so I could only work on it during the weekend. I think I will push a work-in-progress pull request this weekend to get some early feedback.

**Comment by zheli** [comment]
Leave a comment here to confirm that I am still working on this issue, so that gitcoin bot will not remove me from the issue :)

**Comment by vs77bb** [comment]
Hi @zheli this one has returned to an open status on Gitcoin (not sure if on purpose or if it was a mistake)... are you still working on it? I've heard interest from a few people on this so would appreciate an update, if so. Hope you are well!

**Comment by zheli** [comment]
Hi @vs77bb! Thanks for checking out. Yes I am still working on it and trying to finish it up this weekend actually. Perhaps in the future gitcoin should also watch the pull request for updates. I will submit an issue on the gitcoin repo.

Regarding to the issue status on gitcoin, it might be a bug. Who should I contact in regarding to this?

**Comment by carver** [comment]
Thanks @zheli this looks wrapped!

**Comment by gitcoinbot** [comment]
__Work for 0.4 ETH (182.11 USD @ $455.27/ETH) has been submitted by__: 
 1. [@zheli](https://gitcoin.co/profile/zheli) 

 Submitters, please leave a comment to let the funder (@pipermerriam) (and the other parties involved) that you've submitted you work.  If you don't leave a comment, the funder may expire your submission at their discretion. 

 * Learn more [on the gitcoin issue page](https://gitcoin.co/issue/ethereum/web3.py/666)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $2523.12 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by zheli** [comment]
@pipermerriam will I still get paid for this issue? 😄

**Comment by carver** [comment]
Lol, well it definitely shouldn't go to me :P

**Comment by pipermerriam** [comment]
Triggered payment (txn may take some time to confirm)

0x012d386eaa7dd55b272f5facf784b2c17b2a465ba4453c1b8515db6a99c1f665

**Comment by zheli** [comment]
@pipermerriam thx!

**Comment by weberV** [comment]
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__This issue now has a funding of 0.4 ETH (276.84 USD @ $692.1/ETH) attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](http://localhost:8000/issue/ethereum/web3.py/666/92).
 * Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $1,512.27 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)
